### PR TITLE
Draw heatmap

### PR DESCRIPTION
**Pull request by dodget**
This PR resolves ATGU-2166 and ATGU-2247 by:
- Adding a heatmap visualization based on real rubric data coming from canvas
- Refactoring the frontend app to a pattern using `useReducer` and `useContext`

**Notes**
1. HeatMap.js is intended to be refactored out into the `rubric-io` node package
2. A fourth value (denormalized_data) has been added to the JSON repsonse in `rubric_data/views.py`. This is to provide data to the frontend that can easily be transformed into any form that we need. Thanks to Jeremy Guillette for the assist!
3. I went in the direction of implementing `useReducer` at the application level, and passing state to components via `useContext`. This allows us to define a centralized application state and reducer (_controller_). From this point forward, I don't foresee a reason for an individual component to maintain its own individual state.

### CODE REVIEW

**Comment by dodget** [comment]
Thanks for the great review @arthurian! I don't have real answers to some of the general questions, and I hope we can come to a consensus on how the app should be structured.

I like your idea of moving from setter style dispatches to event-based. The way I found myself writing dispatch events was based mostly on how I had defined initial state, and some of the limitations I put on myself in that regard. So I'll start with describing how I was imagining state to be structured:
I was imagining the app state to be a reflection (or model) of the app UI. That's why you see a  `compareAssignments` object in state; it is a reflection of the tab in the UI. And within the `compareAssignments` objects, you'll find `selectors` which lists all of the select fields in the tab. In my mind, there was some cognitive benefit to having a 1:1 relationship between state and UI. However, now that I am considering it, that is inherently inflexible. I like the way you listed the categories of state, and would suggest we flatten `initialState` to something closely resembling that list.
I think that flattening of state would help with reducing the verbosity of the reducer and move to an event-based dispatch where
`dispatch({ type: `compareAssignments-showingRubrics-setValue`, value: '...' })` could become =>
`dispatch({ type: `rubricSelected`, tab: "tabName", value: '...' });`

By adding `tab: "..."` I am trying to make sure that users only control the piece of UI that they are currently on, and not unknowingly updating inactive tabs. However, now that I think about it, perhaps that could be controlled by an `activeTab` value in state...

Long story short, I propose we flatten state based on the categories of state you listed above, and move to an event-based reducer. Since all of the work in this PR was shoe-horned into ATGU-2166 (Heat map visualization) maybe that work could be a new story?

File: frontend/src/components/CompareAssignmentsTab/CompareAssignmentsTab.js [hunk 1]
```diff
@@ -1,46 +1,91 @@
...
-    </Flex>
-    <div className="filter-bar">
-      <Flex>
+import { AppContext } from '../AppState';
+import { pivotHeatMapData } from '../utils';
+
+const CompareAssignmentsTab = () => {
+  const { state, dispatch } = useContext(AppContext);
+
+  // TODO: these transformations and updates to state need to be re-evaluated
```
[the remark this answers is not in the document]
Comment (dodget) [reply frontend/src/components/CompareAssignmentsTab/CompareAssignmentsTab.js:13]: 1. We could do the transformation in the reducer when the payload is received, but I was gravitating toward on-demand computation. If the user never visits  the "Compare Assignments" tab, then why should we bother pivoting the data? Having the pivot live in a dispatch in a component make it so it will only happen _if_ this component is ever rendered. But you are right, that data structure may be useful for other visualizations, so perhaps this should be generalized.

2. I suppose it doesn't need to be stored. I was storing it for performance purposes -- I imagine checking the `.length` of a stored array would be quicker than checking the `.length` of a computed array? I like the idea of leveraging `useMemo` to store the computed value based on equality of inputs, but wouldn't that be functionally the same as storing the value in state? For the sake of future a maintainer's cognitive load, I'd lean towards explicitly defining the value placeholder `initialState` and storing the computed value there, but I'm interested in exploring the `useMemo` route.

File: frontend/src/components/utils.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        return acc;
+      }, 0),
+      dataPoints: assignment.rubric.map((criterion) => criterion.ratings.map((rating) => {
+        const dataPoint = {
+          criterionId: criterion.id,
+          criterion: criterion.description,
+          ratingDescription: rating.description,
+          maxPoints: rating.points,
+          count: payload.denormalized_data.reduce((acc, curr) => (
+            acc + (curr.criterion_id === criterion.id
```
[the remark this answers is not in the document]
Comment (dodget) [reply frontend/src/components/utils.js:23]: I was concerned about that, and `Number()` is a good solution. Thanks!

Yes, so far, this is the only consumer of `denormalized_data`. But that is likely to change.

File: rubric_data/views.py [hunk 3]
```diff
@@ -15,31 +15,89 @@
...
+        'submissions': submissions,
         'students': students,
     }
+    denormalized_data = denormalize(payload)
+    payload['denormalized_data'] = denormalized_data
+    
+    return JsonResponse(payload)
+    
+    
+def denormalize(data):
```
[the remark this answers is not in the document]
Comment (dodget) [reply rubric_data/views.py:48]: True.

File: frontend/src/components/HeatMap/HeatMap.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import * as d3 from 'd3';
+
+export const drawFlatHeatMap = () => 0;
+
+export const drawHeatMap = (id, data) => {
```
[the remark this answers is not in the document]
Comment (dodget) [reply frontend/src/components/HeatMap/HeatMap.js:5]: Lacking documentation seems to be a pattern, eh?

**Comment by arthurian** [comment]
@dodget I completely agree that any larger refactoring deserves its own story. This one does a ton of heavy lifting bootstrapping the structure and setting up the visualization, so it would be good to isolate any larger changes separately.

The point about the app state mapping directly to the UI state makes a lot of sense, and is probably the best way to start with this.  The simplicity of the 1:1 mapping makes it easy to reason about. Having the benefit of seeing what's already there, I was trying to envision what the app state would look like with a couple more tabs, each with nested state, and that's what started to make me wonder if there were any alternatives. 

On the memo side of things, you're probably right that it would be wise to stay away from that and either just call the function to get the derived value, or put it in state to cache it explicitly.
